designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/provider/ProcessingTypeDataProvider.scala

```diff
@@ -1,56 +1,77 @@
 package pl.touk.nussknacker.ui.process.processingtype.provider
 
+import cats.effect.{IO, Ref}
+import cats.effect.std.Mutex
+import cats.effect.unsafe.IORuntime
+import cats.implicits.toTraverseOps
+import com.typesafe.scalalogging.LazyLogging
 import pl.touk.nussknacker.engine.api.process.ProcessingType
 import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
 import pl.touk.nussknacker.security.Permission
 import pl.touk.nussknacker.ui.UnauthorizedError
 import pl.touk.nussknacker.ui.process.processingtype.ValueWithRestriction
 import pl.touk.nussknacker.ui.security.api.LoggedUser
 
-import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.CopyOnWriteArrayList
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Try}
 
 /**
   * ProcessingType is a context of application. One ProcessingType can't see data from another ProcessingType.
   * Services run inside one ProcessingType scope behave differently from services run in another scope.
   *
   * This class is meant to provide access to some scope of data inside context of application to the user.
   * We don't want to pass all ProcessingType's data to every service because it would complicate testing of services
-  * and would broke isolation between areas of application. Due to that, this class is a `Functor`
+  * and would brake isolation between areas of application. Due to that, this class is a `Functor`
   * (to be precise `BiFunctor` but more on that below) which allows to transform the scope of `Data`.
   *
   * Sometimes it is necessary to have access also to combination of data across all ProcessingTypes. Due to that
   * this class is a `BiFunctor` which second value named as `CombinedData`
   *
   * This class caches `Data` and `CombinedData` wrapped in `ProcessingTypeDataState` to avoid computations of
-  * transformations during each lookup to `Data`/`CombinedData`. It behave similar to `Observable` where given
-  * transformed `ProcessingTypeDataProvider` check its parent if `ProcessingTypeDataState.stateIdentity` changed.
+  * transformations during each lookup to `Data`/`CombinedData`.
   *
   * ProcessingType is associated with Category e.g. Fraud Detection, Marketing. Given user has access to certain
-  * categories see `LoggedUser.can`. Due to that, during each access to `Data`, user is authorized if he/she
-  * has access to category.
+  * categories see `LoggedUser.can`. Due to that, during each access to `Data`, user is authorized if they
+  * have access to category.
   */
-trait ProcessingTypeDataProvider[+Data, +CombinedData] {
+abstract class ProcessingTypeDataProvider[Data, CombinedData](
+    initialState: ProcessingTypeDataState[Data, CombinedData]
+)(implicit ioRuntime: IORuntime)
```

Comment:
Why don't we create this IORuntime in the constructor of this class? The child class doesn't need it. The runtime is only needed by this class's internals. Moreover, the IORuntime is not tweaked or configured. The lifecycle of this object can be related to the lifecycle of this class and managed from the class's internals.